src/app/campaign/store/effects/campaign.effects.ts

```diff
@@ -25,34 +25,124 @@ export class CampaignEffects {
 
   @Effect()
   campaignFormSave$ = this.actions$.pipe(
-    ofType(CampaignActionTypes.CAMPAIGN_FORM_SAVE),
-    map((action: actions.CampaignFormSave) => action.payload),
-    mergeMap((payload: { campaign }) => {
-      const result = payload.campaign.id
-        ? this.campaignService.updateCampaign(payload.campaign)
-        : this.campaignService.createCampaign(payload.campaign);
-      return result.pipe(
-        tap(({ campaignDoc }: { campaignDoc: HalDoc }) => {
-          // TODO: incorporate flight changes and the rest of routing here
-          // navigate to newly created campaign/flight
-          // const flightId = this.router.url.split('/flight/').pop();
-          // if (this.router.url.includes('/flight/') && !changes.flights[flightId]) {
-          //   // flight not in changes (flight deleted)
-          //   this.router.navigate(['/campaign', campaign.id]);
-          // } else if (this.router.url.includes('/flight/') && flightId !== changes.flights[flightId]) {
-          //   // flight id changed (flight created)
-          //   this.router.navigate(['/campaign', campaign.id, 'flight', changes.flights[flightId]]);
-          // } else
-          if (payload.campaign.id !== campaignDoc.id) {
-            // campaign id changed (campaign created)
-            this.router.navigate(['/campaign', campaignDoc.id]);
-          }
+    ofType(ActionTypes.CAMPAIGN_SAVE),
+    map((action: actions.CampaignSave) => action.payload),
+    mergeMap((payload: CampaignFormSave) => {
```

Comment:
Save the campaign, then delete any softDeleted flights, update any changed flights, and create and new flights. Return results as HalDoc with flight ids mapped to their results.

Reply:
forkJoin will not return a result if any of the Observables has an error (or completes without emitting.) The error would bubble up into the stream and end up in this effect's catchError. The effects has to return an Observable of an action even in the event of error in order to keep the stream moving or the application itself would error with something like `sttream expected`. So we return a failure action. The error is saved onto the campaign state in teh campaign reducer, then it would be displayed through the error selector.